File: ruoyi-common/src/test/java/com/ruoyi/common/utils/PasswordPolicyTest.java
package com.ruoyi.common.utils;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.List;
import org.junit.jupiter.api.Test;

class PasswordPolicyTest
{
    @Test
    void accepts_a_compliant_password()
    {
        assertThat(PasswordPolicy.violation("Abcdef1!", null)).isEmpty();
    }

    @Test
    void flags_each_rule_in_order()
    {
        assertThat(PasswordPolicy.violation("Ab1!", null)).contains("password.tooShort");
        assertThat(PasswordPolicy.violation("A".repeat(20) + "b1!aaaaaaaaaaaaa", null)).contains("password.tooLong");
        assertThat(PasswordPolicy.violation("abcdef1!", null)).contains("password.needUpper");
        assertThat(PasswordPolicy.violation("ABCDEF1!", null)).contains("password.needLower");
        assertThat(PasswordPolicy.violation("Abcdefg!", null)).contains("password.needDigit");
        assertThat(PasswordPolicy.violation("Abcdefg1", null)).contains("password.needSpecial");
    }

    @Test
    void rejects_reuse_of_the_current_password()
    {
        String encoded = SecurityUtils.encryptPassword("Abcdef1!");
        assertThat(PasswordPolicy.violation("Abcdef1!", encoded)).contains("password.sameAsCurrent");
    }

    @Test
    void a_different_new_password_is_allowed_against_a_current_one()
    {
        String encoded = SecurityUtils.encryptPassword("Abcdef1!");
        assertThat(PasswordPolicy.violation("Zxcvbn2@", encoded)).isEmpty();
    }

    @Test
    void rejects_a_password_that_contains_a_personal_term()
    {
        assertThat(PasswordPolicy.violation("Lovelace1!", null, List.of("Ada", "Lovelace", "ada")))
                .contains("password.noPersonal");
        // short terms (< 3 chars) are ignored
        assertThat(PasswordPolicy.violation("Abcdef1!", null, List.of("Ab"))).isEmpty();
        // a clean password with personal terms supplied still passes
        assertThat(PasswordPolicy.violation("Zxcvbn2@", null, List.of("Ada", "Lovelace"))).isEmpty();
    }
}

File: nadoumi-web/tests/unit/passwordPolicy.test.ts
import { describe, it, expect } from 'vitest'
import { passwordPolicyKey, passwordChecks } from '~/utils/passwordPolicy'

describe('passwordPolicyKey', () => {
  it('accepts a compliant password', () => {
    expect(passwordPolicyKey('Abcdef1!')).toBeNull()
  })

  it('flags each rule in order', () => {
    expect(passwordPolicyKey('Ab1!')).toBe('validation.password.tooShort')
    expect(passwordPolicyKey('a'.repeat(33) + 'B1!')).toBe('validation.password.tooLong')
    expect(passwordPolicyKey('abcdef1!')).toBe('validation.password.needUpper')
    expect(passwordPolicyKey('ABCDEF1!')).toBe('validation.password.needLower')
    expect(passwordPolicyKey('Abcdefg!')).toBe('validation.password.needDigit')
    expect(passwordPolicyKey('Abcdefg1')).toBe('validation.password.needSpecial')
  })

  it('rejects reuse of the current password', () => {
    expect(passwordPolicyKey('Abcdef1!', 'Abcdef1!')).toBe('validation.password.sameAsCurrent')
  })
})

describe('passwordChecks', () => {
  it('reports a strong password with all rules passing', () => {
    const r = passwordChecks('Abcdef1!')
    expect(r.strong).toBe(true)
    expect(r.firstError).toBeNull()
    expect(r.rules.every(x => x.ok)).toBe(true)
  })

  it('adds a personal-term rule only when forbidden terms are supplied', () => {
    expect(passwordChecks('Abcdef1!').rules.some(x => x.key === 'validation.password.noPersonal')).toBe(false)
    const r = passwordChecks('Lovelace1!', { forbidden: ['Ada', 'Lovelace'] })
    expect(r.strong).toBe(false)
    expect(r.firstError).toBe('validation.password.noPersonal')
  })

  it('reports a confirm mismatch last', () => {
    const r = passwordChecks('Abcdef1!', { confirm: 'Abcdef2@' })
    expect(r.strong).toBe(true) // policy itself is fine
    expect(r.firstError).toBe('validation.password.mismatch')
  })

  it('short forbidden terms (<3 chars) are ignored', () => {
    expect(passwordChecks('Abcdef1!', { forbidden: ['Ab'] }).firstError).toBeNull()
  })
})
